tiledb/sm/query/readers/sparse_index_reader_base.h

```diff
@@ -296,11 +296,72 @@ class SparseIndexReaderBase : public ReaderBase {
    * it is really required to determine if a query is incomplete from the client
    * side of a cloud request.
    */
-  struct ReadState {
+  class ReadState {
+   public:
+    /* ********************************* */
+    /*     CONSTRUCTORS & DESTRUCTORS    */
+    /* ********************************* */
+
+    ReadState() = delete;
+
+    ReadState(size_t frag_idxs_len)
+        : frag_idx_(frag_idxs_len) {
+    }
+
+    ReadState(std::vector<FragIdx>&& frag_idx, bool done_adding_result_tiles)
+        : frag_idx_(std::move(frag_idx))
+        , done_adding_result_tiles_(done_adding_result_tiles) {
+    }
+
+    /* ********************************* */
+    /*                API                */
+    /* ********************************* */
+
+    /**
+     * Return whether the tiles that will be processed are loaded in memory.
+     * @return Done adding result tiles.
+     */
+    const bool& done_adding_result_tiles() const {
+      return done_adding_result_tiles_;
+    }
+
+    /**
+     * Sets the flag that determines whether the tiles that will be processed
+     * are loaded in memory.
+     * @param done_adding_result_tiles Done adding result tiles.
+     * @return void
+     */
+    inline void set_done_adding_result_tiles(bool done_adding_result_tiles) {
+      done_adding_result_tiles_ = done_adding_result_tiles;
+    }
+
+    /**
+     * Sets a value in the fragment index vector.
+     * @param idx The index of the vector.
+     * @param val The value to set frag_idx[idx] to.
+     * @return void
+     */
+    inline void set_frag_idx(uint64_t idx, FragIdx val) {
+      if (idx >= frag_idx_.size()) {
+        throw std::runtime_error(
+            "ReadState::set_frag_idx: idx greater than frag_idx_'s size.");
+      }
+      frag_idx_[idx] = std::move(val);
+    }
+
+    /**
+     * Returns a read-only version of the fragment index vector.
+     * @return The fragment index vector.
+     */
+    const std::vector<FragIdx>& frag_idx() const {
+      return frag_idx_;
+    }
+
+   private:
```

Comment:
We could add a private attribute comment box here.

Reply:
Done!